hax/hax/motr/__init__.py

```diff
@@ -148,52 +162,93 @@ def _entrypoint_request_cb(self, reply_context: Any, req_id: Any,
 
     def send_entrypoint_request_reply(self, message: EntrypointRequest):
         reply_context = message.reply_context
         req_id = message.req_id
         remote_rpc_endpoint = message.remote_rpc_endpoint
         process_fid = message.process_fid
+        e_rc = EAGAIN
 
         LOG.debug('Processing entrypoint request from remote endpoint'
                   " '{}', process fid {}".format(remote_rpc_endpoint,
                                                  str(process_fid)))
         sess = principal_rm = confds = None
         try:
             util = self.consul_util
-            sess = util.get_leader_session_no_wait()
-            principal_rm = util.get_session_node(sess)
-            confds = util.get_confd_list()
-            rm_fid = util.get_rm_fid()
+            # When stopping, there's a possibility that hax may receive
+            # an entrypoint request from motr land. In order to unblock
+            # motr land, reply with entrypoint request with no confds
+            # and RM endpoints as the processes might have already
+            # stopped.
+            rc_quorum = 0
+            if self.is_stopping:
+                confds = []
+                rm_fid = Fid(0, 0)
+            else:
+                sess = util.get_leader_session_no_wait()
+                principal_rm = util.get_session_node(sess)
+                confds = util.get_confd_list()
+                rm_fid = util.get_rm_fid()
+
+            # Hax may receive entrypoint requests multiple times during its
+            # lifetime. Hax starts motr rconfc to invoke spiel commands. Motr
+            # rconfc establishes connection with principal RM, in case of
+            # principal RM failure, rconfc invalidates its confc and again
+            # requests entrypoint in a hope that there will be another confd
+            # and principal RM elected so that rconfc can resume its
+            # functionality. During shutdown, when each motr process stops,
+            # including confds, hax broadcasts M0_NC_FAILED event for every
+            # STOPPED or FAILED motr process. Motr rconfc on receiving the
+            # failed events for confds, goes re-requests entrypoint information
+            # and this goes on in a loop. In order to break this loop, the
+            # the entrypoint reply must only report alive confds and rm
+            # endpoints. While doing this we need to handle the bootstrapping
+            # case, so we wait until bootstrapping is done that is all the
+            # motr services are up and then account for the motr process
+            # failures and lookup this list of failed motr processes before
+            # replying to entrypoint request.
+            if self.spiel_ready:
+                active_confds = confds
+                for confd in active_confds:
+                    with util.failed_svcs_lock:
+                        if confd.fid in util.failed_svcs:
+                            confds.remove(confd)
```

Comment:
I think there is no need to exclude failed confd's, this may impact quorum of confd's. Rconfc logic manages it internally.

Reply:
With non-empty entrypoint reply I have observed that rconfc stop blocks and hax keeps getting entrypoint requests from motr land.

Reply:
see,
```
May 20 16:35:54 ssc-vm-1623.colo.seagate.com hare-hax[118893]: motr[118893]:  4900   WARN  [ha/halon/interface.c:1073:m0_halon_interface_entrypoint_reply]  req_id=0:14 rc=0 confd_nr=3 confd_quorum=2 rm_fid=<7300000000000001:a> rm_eps=10.230.250.51@tcp:12345:2:1
May 20 16:35:54 ssc-vm-1623.colo.seagate.com hare-hax[118893]: motr[118893]:  4540   WARN  [ha/halon/interface.c:490:halon_interface_link_reused_cb]  req_id=0:14 hl=0x7fd7d4009360 ep=10.230.250.51@tcp:12345:1:1
May 20 16:35:54 ssc-vm-1623.colo.seagate.com hare-hax[118893]: 2021-05-20 16:35:54,265 [DEBUG] {Dummy-1} Entrypoint request has been replied to
May 20 16:35:55 ssc-vm-1623.colo.seagate.com hare-hax[118893]: motr[118893]:  4a20  ERROR  [conf/confc.c:557:m0_confc_init_wait]  <! rc=-110 confc=0x7fd86404cfa0 confd_addr=10.230.250.51@tcp:12345:2:1
May 20 16:35:56 ssc-vm-1623.colo.seagate.com hare-hax[118893]: motr[118893]:  4a20  ERROR  [conf/confc.c:557:m0_confc_init_wait]  <! rc=-110 confc=0x7fd86404fef0 confd_addr=10.230.250.53@tcp:12345:2:1
May 20 16:35:57 ssc-vm-1623.colo.seagate.com hare-hax[118893]: motr[118893]:  4a20  ERROR  [conf/confc.c:557:m0_confc_init_wait]  <! rc=-110 confc=0x7fd864051630 confd_addr=10.230.250.48@tcp:12345:2:1
May 20 16:35:57 ssc-vm-1623.colo.seagate.com hare-hax[118893]: motr[118893]:  4b30  ERROR  [conf/rconfc.c:1699:rconfc_herd_update]  <! rc=-2
May 20 16:35:57 ssc-vm-1623.colo.seagate.com hare-hax[118893]: motr[118893]:  4b30  ERROR  [conf/rconfc.c:1948:rconfc_entrypoint_consume]  <! rc=-2
May 20 16:35:57 ssc-vm-1623.colo.seagate.com hare-hax[118893]: motr[118893]:  4b30  ERROR  [conf/rconfc.c:1968:rconfc_start_internal]  <! rc=-2
May 20 16:35:57 ssc-vm-1623.colo.seagate.com hare-hax[118893]: motr[118893]:  5bb0   WARN  [ha/halon/interface.c:366:halon_interface_entrypoint_request_cb]  req_id=0:15 remote_rpc_endpoint=10.230.250.51@tcp:12345:1:1 process_fid=<7200000000000001:6> git_rev_id=51d02dac8c2096fbe82bd551dc0895d8967d0212 generation=18 pid=118893
May 20 16:35:57 ssc-vm-1623.colo.seagate.com hare-hax[118893]: 2021-05-20 16:35:57,273 [DEBUG] {Dummy-1} Received entrypoint request from remote endpoint '10.230.250.51@tcp:12345:1:1', process fid = 0x7200000000000001:0x6 The request will be processed in another thread.
May 20 16:35:57 ssc-vm-1623.colo.seagate.com hare-hax[118893]: 2021-05-20 16:35:57,273 [DEBUG] {Dummy-1} enqueue entrypoint request for 10.230.250.51@tcp:12345:1:1
May 20 16:35:57 ssc-vm-1623.colo.seagate.com hare-hax[118893]: 2021-05-20 16:35:57,274 [DEBUG] {Dummy-1} Processing entrypoint request from remote endpoint '10.230.250.51@tcp:12345:1:1', process fid 0x7200000000000001:0x6
May 20 16:35:57 ssc-vm-1623.colo.seagate.com hare-hax[118893]: 2021-05-20 16:35:57,288 [DEBUG] {Dummy-1} Passing the entrypoint reply to hax.c layer
May 20 16:35:57 ssc-vm-1623.colo.seagate.com hare-hax[118893]: motr[118893]:  4900   WARN  [ha/halon/interface.c:1073:m0_halon_interface_entrypoint_reply]  req_id=0:15 rc=0 confd_nr=3 confd_quorum=2 rm_fid=<7300000000000001:a> rm_eps=10.230.250.51@tcp:12345:2:1
```
Hax motr land keeps sending the entrypoint requests, other remaining confd's shutdown is also delayed.
```
May 20 16:37:37 ssc-vm-1623.colo.seagate.com motr-server[119434]: motr[119434]:  95e0   WARN  [reqh/reqh.c:561:m0_reqh_fop_handle]  fop "Rpc conn establish"@0x7ff2480a2110 disallowed: -111.
May 20 16:37:37 ssc-vm-1623.colo.seagate.com motr-server[119434]: motr[119434]:  95e0  ERROR  [reqh/reqh.c:568:m0_reqh_fop_handle]  <! rc=-108 Service shutdown.
May 20 16:37:40 ssc-vm-1623.colo.seagate.com motr-server[119434]: motr[119434]:  94e0  ERROR  [reqh/reqh.c:454:m0_reqh_fop_allow]  <! rc=-111
May 20 16:37:40 ssc-vm-1623.colo.seagate.com motr-server[119434]: motr[119434]:  95e0   WARN  [reqh/reqh.c:561:m0_reqh_fop_handle]  fop "Rpc conn establish"@0x7ff2480a2110 disallowed: -111.
May 20 16:37:40 ssc-vm-1623.colo.seagate.com motr-server[119434]: motr[119434]:  95e0  ERROR  [reqh/reqh.c:568:m0_reqh_fop_handle]  <! rc=-108 Service shutdown.
May 20 16:37:43 ssc-vm-1623.colo.seagate.com motr-server[119434]: motr[119434]:  94e0  ERROR  [reqh/reqh.c:454:m0_reqh_fop_allow]  <! rc=-111
May 20 16:37:43 ssc-vm-1623.colo.seagate.com motr-server[119434]: motr[119434]:  95e0   WARN  [reqh/reqh.c:561:m0_reqh_fop_handle]  fop "Rpc conn establish"@0x7ff2480a2110 disallowed: -111.
May 20 16:37:43 ssc-vm-1623.colo.seagate.com motr-server[119434]: motr[119434]:  95e0  ERROR  [reqh/reqh.c:568:m0_reqh_fop_handle]  <! rc=-108 Service shutdown.
May 20 16:37:46 ssc-vm-1623.colo.seagate.com motr-server[119434]: motr[119434]:  94e0  ERROR  [reqh/reqh.c:454:m0_reqh_fop_allow]  <! rc=-111
May 20 16:37:46 ssc-vm-1623.colo.seagate.com motr-server[119434]: motr[119434]:  95e0   WARN  [reqh/reqh.c:561:m0_reqh_fop_handle]  fop "Rpc conn establish"@0x7ff2480a2110 disallowed: -111.
May 20 16:37:46 ssc-vm-1623.colo.seagate.com motr-server[119434]: motr[119434]:  95e0  ERROR  [reqh/reqh.c:568:m0_reqh_fop_handle]  <! rc=-108 Service shutdown
```
and one of the confd, that is stopping keeps sending RM_BORROW requests,
```
771888  15640197.087.229544  97f9860  DEBUG   rpc/item.c:737:m0_rpc_item_change_state             0x7ff248009c10 [REQUEST/M0_RM_FOP_BORR771889  15640197.087.231982  97f98e0  CALL    rpc/rpc_machine.c:903:item_received                 <
771890  15640197.087.232276  97f9880  CALL    fop/fop.c:182:m0_fop_put                            > fop: 0x7ff248009bf0 Credit Borrow [1771891  15640197.087.233470  97f9850  CALL    fop/fop.c:152:m0_fop_release                        >
771892  15640197.087.234294  97f97f0  CALL    fop/fop.c:139:m0_fop_fini                           > fop: 0x7ff248009bf0 Credit Borrow
771893  15640197.087.235762  97f9760  CALL    rpc/item.c:396:m0_rpc_item_fini                     > 0x7ff248009c10 [REQUEST/M0_RM_FOP_BORROW(88)] rc=0
```